glean-core/src/metrics/dual_labeled_counter.rs

```diff
@@ -0,0 +1,375 @@
+            DynamicLabelType::CategoryOnly(_) => {
+                if category_count >= MAX_LABELS || !label_is_valid(category, glean, meta) {
+                    category = OTHER_LABEL;
+                }
+            }
+            DynamicLabelType::KeyAndCategory(_) => {
+                if key_count >= MAX_LABELS || !label_is_valid(key, glean, meta) {
+                    key = OTHER_LABEL;
+                }
+                if category_count >= MAX_LABELS || !label_is_valid(category, glean, meta) {
+                    category = OTHER_LABEL;
+                }
+            }
+        }
+        combine_base_identifier_and_labels(base_identifier, key, category)
+    } else {
+        record_error(
+            glean,
+            meta,
+            ErrorType::InvalidLabel,
+            "Invalid `DualLabeledCounter` label format, unable to determine key and/or category",
+            None,
+        );
+        combine_base_identifier_and_labels(base_identifier, OTHER_LABEL, OTHER_LABEL)
+    }
+}
+
+fn label_is_valid(label: &str, glean: &Glean, meta: &CommonMetricDataInternal) -> bool {
+    if label.len() > MAX_LABEL_LENGTH {
+        let msg = format!(
+            "label length {} exceeds maximum of {}",
+            label.len(),
+            MAX_LABEL_LENGTH
+        );
+        record_error(glean, meta, ErrorType::InvalidLabel, msg, None);
+        false
+    } else if label.chars().any(|c| !c.is_ascii() || c.is_ascii_control()) {
+        let msg = format!("label must be printable ascii, got '{:?}'", label);
+        record_error(glean, meta, ErrorType::InvalidLabel, msg, None);
+        false
+    } else {
+        true
+    }
+}
+
+fn count_keys_and_categories(meta: &CommonMetricDataInternal, glean: &Glean) -> (usize, usize) {
+    let base_identifier = &meta.base_identifier();
+    let prefix = format!("{base_identifier}{RECORD_SEPARATOR}");
+    let mut seen_keys: Vec<String> = Vec::new();
+    let mut seen_categorires: Vec<String> = Vec::new();
+    let mut snapshotter = |mut metric_id: &[u8], _: &Metric| {
+        let mut metric_id_str = String::new();
+        _ = metric_id.read_to_string(&mut metric_id_str);
+
+        // Split full identifier on the ASCII Record Separator (\x1e)
+        let parts: Vec<&str> = metric_id_str.split(RECORD_SEPARATOR).collect();
+
+        if parts.len() == 2 {
+            if !seen_keys.contains(&parts[0].into()) {
+                seen_keys.push(parts[0].into());
+            }
+            if !seen_categorires.contains(&parts[1].into()) {
+                seen_categorires.push(parts[1].into());
+            }
+        } else {
+            println!("DualLabeledCounter label does not contain exactly 2 parts");
+        }
+    };
+
+    let lifetime = meta.inner.lifetime;
+    for store in &meta.inner.send_in_pings {
+        glean
+            .storage()
+            .iter_store_from(lifetime, store, Some(&prefix), &mut snapshotter);
+    }
+
+    (seen_keys.len(), seen_categorires.len())
```

Comment:
A thing I realized here:
This means we count categories, regardless of under which `key` they are.
So if we have

`key: a, categories: [0, 1, ... 14]`
`key: b, categories: [15]`

then `key: c` will _always_ end up recording to `__other__`.
that seems a bit unintuitive.
I assumed it's 16 categories per unique key.
Of course that's harder to implement.

The proposal is vague about this, so I could be convinced the implemented behavior is correct.